pkg/assembler/grammar.go

```diff
@@ -0,0 +1,199 @@
+package assembler
+
+import (
+	"fmt"
+	"math"
+)
+
+// Grammar and AST
+
+type CasmProgram struct {
+	Instructions []Instruction `@@`
+}
+
+type Instruction struct {
+	Core      *CoreInstruction `@@`
+	ApPlusOne bool             `( "," @"ap" "+" "+" )? ";" |`
+	ApPlus    *ApPlus          `@@ ";"`
+}
+
+type CoreInstruction struct {
+	AssertEq *AssertEq `@@ |`
+	Jnz      *Jnz      `@@ |`
+	Jump     *Jump     `@@ |`
+	Call     *Call     `@@ |`
+	Ret      *Ret      `@@ `
+}
+
+type AssertEq struct {
+	Dst   *Deref      `@@`
+	Value *Expression `"=" @@`
+}
+
+type Jump struct {
+	JumpType string      `"jmp" @("rel" | "abs")`
+	Value    *Expression `@@`
+}
+
+type Jnz struct {
+	Value     *DerefOrImm `"jmp" "rel" @@`
+	Condition *Deref      `"if" @@ "!" "=" "0"`
+}
+
+type Call struct {
+	CallType string      `"call" @("rel" | "abs")`
+	Value    *DerefOrImm `@@`
+}
+
+type Ret struct {
+	Ret string `"ret"`
+}
+
+type ApPlus struct {
+	Value *Expression `"ap" "+" "=" @@`
+}
+
+type Expression struct {
+	DoubleDeref   *DoubleDeref   `@@ |`
+	MathOperation *MathOperation `@@ |`
+	Deref         *Deref         `@@ |`
+	Immediate     *string        `@Int`
+}
+
+type Deref struct {
+	Name   string  `"[" @("ap" | "fp")`
+	Offset *Offset `@@? "]"`
+}
+
+type DoubleDeref struct {
+	Deref  *Deref  `"[" @@`
+	Offset *Offset `@@? "]"`
+}
+
+type Offset struct {
+	Sign  string `@("+" | "-")`
+	Value *int   `@Int`
+}
+
+type MathOperation struct {
+	Lhs      *Deref      `@@`
+	Operator string      `@("+" | "*")`
+	Rhs      *DerefOrImm `@@`
+}
+
+type DerefOrImm struct {
+	Deref     *Deref  `@@ |`
+	Immediate *string `@Int`
+}
+
+// AST Functionality
+
+type CoreInstructioner interface {
+	Expression() Expressioner
+}
+
+func (instruction Instruction) Unwrap() CoreInstructioner {
+	if instruction.ApPlus != nil {
+		return instruction.ApPlus
+	} else if instruction.Core.AssertEq != nil {
+		return instruction.Core.AssertEq
+	} else if instruction.Core.Jump != nil {
+		return instruction.Core.Jump
+	} else if instruction.Core.Jnz != nil {
+		return instruction.Core.Jnz
+	} else if instruction.Core.Call != nil {
+		return instruction.Core.Call
+	} else {
+		return instruction.Core.Ret
+	}
+}
```

Comment:
Since all fields are exported (rightfully so imho), I don't think we gain much from the `CoreInstructioner` and `Expressioner` interfaces. If we remove them and the associated methods, we get to eliminate a lot of code with only minor refactors in `assembler.go`.

I tried it out a bit locally and it seemed to work. Let me know if I'm missing something.

Reply:
This also allows us to simplify other things, like removing the `ApPlus` struct, which is nice.

Reply:
Yes, we can definitely get rid of `CoreInstructioner`. I initially implemented it thinking I would use it a lot more. 

The purpose of `Expressioner` is to have both expression rules `Expression` and `DerefOrImmediate` as the same type. I think we can remove it as well by putting `DerefOrImmediate` as a child of `Expression` (while removing its `Deref` and `Immediate` children. I am ok with whatever solution, the current implemented, or an updated one.

The purpose of these interfaces was to hide away all of the pattern matching of the ast, since the assembler shouldn't really care about. I wouldn't like it to have it implemented in the assembly code

Reply:
> DerefOrImmediate as a child of Expression (while removing its Deref and Immediate children.

This sounds interesting. The `As<Struct>` interface methods don't feel quite right to me, but maybe we can leave this comment unresolved for now while we discuss the other items. I need a moment to think about it.

Reply:
After further talks, we decided to leave as is for now and perhaps improve it later